protocols/mdns/src/behaviour.rs

```diff
@@ -36,29 +36,53 @@ use smallvec::SmallVec;
 use std::collections::hash_map::{Entry, HashMap};
 use std::{cmp, fmt, io, net::IpAddr, pin::Pin, task::Context, task::Poll, time::Instant};
 
-#[cfg(feature = "async-io")]
-use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+/// A Fallible Default trait. Used on [`GenMdns`] to Allow accepting `IfWatcher`'s.
+pub trait TryDefault: Sized {
+    /// Fallible equivalent of [`Default::default`].
+    fn try_default() -> Result<Self, std::io::Error>;
+}
 
 /// The type of a [`GenMdns`] using the `async-io` implementation.
 #[cfg(feature = "async-io")]
-pub type Mdns = GenMdns<AsyncUdpSocket, AsyncTimer>;
+pub mod async_io {
+    use super::{GenMdns, TryDefault};
+    use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+    use if_watch::smol::IfWatcher;
 
-#[cfg(feature = "tokio")]
-use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    pub type Mdns = GenMdns<AsyncUdpSocket, AsyncTimer, IfWatcher>;
+
+    impl TryDefault for IfWatcher {
+        fn try_default() -> Result<Self, std::io::Error> {
+            Self::new()
+        }
+    }
+}
 
 /// The type of a [`GenMdns`] using the `tokio` implementation.
 #[cfg(feature = "tokio")]
-pub type TokioMdns = GenMdns<TokioUdpSocket, TokioTimer>;
+pub mod tokio {
+    use super::{GenMdns, TryDefault};
+    use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    use if_watch::tokio::IfWatcher;
+
+    pub type Mdns = GenMdns<TokioUdpSocket, TokioTimer, IfWatcher>;
+
+    impl TryDefault for IfWatcher {
+        fn try_default() -> Result<Self, std::io::Error> {
+            Self::new()
+        }
+    }
```

Comment:
An alternative way of doing this which does not require a trait would be to say:

```rust
impl GenMdns<TokioUdpSocket, TokioTimer, IfWatcher> {
    /// Builds a new `Mdns` behaviour.
    pub fn new(config: MdnsConfig) -> io::Result<Self> {
        Ok(Self {
            config,
            if_watch: IfWatcher::new(),
            iface_states: Default::default(),
            discovered_nodes: Default::default(),
            closest_expiration: Default::default(),
        })
	}
}
```

Reply:
Untested but I think this should work!

Reply:
Yeah, that works if we remove `GenMdns::new` otherwise compiler complains regarding duplication. The disadvantage I see with that is that then methods are scattered between `{tokio, async_io}::Mdns` and `GenMdns`, would you prefer that approach? Tbh ideally I think it would be great if we could deprecate/hide `GenMdns` and just present `{tokio, async_io}::Mdns` to the end user.

Reply:
We had a similar discussion in `libp2p-tcp`: https://github.com/libp2p/rust-libp2p/pull/2961#pullrequestreview-1129815307

From what I remember, the functions don't show up for the type alias at all so probably best to not do this at all.

The need for another trait is a bit annoying though.